### PR TITLE
Match string

### PR DESCRIPTION
**Pull request by NicolasAubry**

### CODE REVIEW

File: tachyon/core/heuristics/logbehaviorchange.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# Tachyon - Fast Multi-Threaded Web Discovery Tool
```
Comment (NicolasAubry) [inline tachyon/core/heuristics/logbehaviorchange.py:1]: Nothing new here, just a class moved to its own file.

File: test/logbehaviorchange_test.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
+# Tachyon - Fast Multi-Threaded Web Discovery Tool
```
Comment (NicolasAubry) [inline test/logbehaviorchange_test.py:1]: Also not new code, just a test case moved to its own file.

File: test/filefetcher_test.py [hunk 3]
```diff
@@ -22,11 +22,13 @@
...
 from hammertime.ruleset import StopRequest, RejectRequest
+from hammertime.kb import KnowledgeBase
 
 from tachyon.core.filefetcher import FileFetcher
 from fixtures import async, create_json_data, FakeHammerTimeEngine, SetResponseCode, fake_future, SetResponseContent,\
-    RaiseForPaths
-from tachyon.core import database
+    RaiseForPaths, SetFlagInResult
+from tachyon.core import database, conf
+from tachyon.__main__ import setup_hammertime_heuristics
```
Comment (lphuberdeau) [inline test/filefetcher_test.py]: Strange import, setup_hammertime_heuristics probably should not be in there.

File: tachyon/core/filefetcher.py [hunk 4]
```diff
@@ -41,6 +41,8 @@ def __init__(self, host, hammertime):
         for future in asyncio.as_completed(requests):
             try:
                 entry = await future
+                if self._is_entry_invalid(entry):
+                    raise RejectRequest("Response is a soft404")
```
Comment (lphuberdeau) [inline tachyon/core/filefetcher.py]: Raising only to catch in the same scope. Use continue instead.

File: tachyon/core/heuristics/matchstring.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+    async def after_response(self, entry):
+        if "file" in entry.arguments:
+            if "match_string" in entry.arguments["file"]:
+                string = entry.arguments["file"]["match_string"]
+                entry.result.string_match = string in entry.response.content
+            elif "match_bytes" in entry.arguments["file"]:
+                raw_data = entry.response.raw
+                raw_string = entry.arguments["file"]["match_bytes"].encode("utf-8")
+                entry.result.string_match = raw_string in binascii.hexlify(raw_data)
```
Comment (lphuberdeau) [inline tachyon/core/heuristics/matchstring.py]: Should not convert the whole body in a different format. Convert the needle instead of the haystack.

**Review by lphuberdeau — CHANGES_REQUESTED** [review]

File: test/filefetcher_test.py (hunk 3 above)
Comment (NicolasAubry) [reply test/filefetcher_test.py]: It there so the same heuristics are used both in the tests and the production. I think it is more maintainable than duplicating the heuristics setup in the test case, as adding a heuristic that affect the behavior would immediately cause the tests to fail.

Comment (lphuberdeau) [reply test/filefetcher_test.py]: The issue is not about re-using, just about the location of the code being reused.

**Review by lphuberdeau — APPROVED** [review]

**Pull request merged**